common/dal2/src/services/s3.rs

```diff
@@ -12,64 +12,202 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use async_compat::CompatExt;
+use std::pin::Pin;
+use std::str::FromStr;
+use std::task::Context;
+use std::task::Poll;
+
 use async_trait::async_trait;
-use rusoto_core::Region;
-use rusoto_s3::GetObjectRequest;
-use rusoto_s3::S3Client;
-use rusoto_s3::S3 as RusotoS3;
+use aws_sdk_s3 as AwsS3;
+use futures::TryStreamExt;
 
+use crate::credential::Credential;
+use crate::error::Error;
 use crate::error::Result;
 use crate::ops::Read;
 use crate::ops::ReadBuilder;
 use crate::ops::Reader;
 
-/// TODO: https://github.com/datafuselabs/databend/issues/3677
-#[derive(Default)]
+#[derive(Default, Debug, Clone)]
 pub struct Builder {
+    pub root: Option<String>,
+
     pub bucket: String,
-    pub endpoint: String,
-    pub credential: String,
+    pub credential: Option<Credential>,
+    pub endpoint: Option<String>,
+
+    /// disable_ssl controls whether to use SSL when connecting.
+    pub disable_ssl: bool,
+    /// enable_path_style controls whether to use path style or virtual host style.
+    pub enable_path_style: bool,
 }
 
 impl Builder {
-    pub fn finish(self) -> Backend {
-        Backend {
-            client: S3Client::new(Region::default()),
-            bucket: self.bucket,
+    pub fn root(&mut self, root: &str) -> &mut Self {
+        self.root = Some(root.to_string());
+
+        self
+    }
+
+    pub fn bucket(&mut self, bucket: &str) -> &mut Self {
+        self.bucket = bucket.to_string();
+
+        self
+    }
+
+    pub fn credential(&mut self, credential: Credential) -> &mut Self {
+        self.credential = Some(credential);
+
+        self
+    }
+
+    pub fn endpoint(&mut self, endpoint: &str) -> &mut Self {
+        self.endpoint = Some(endpoint.to_string());
+
+        self
+    }
+
+    pub fn disable_ssl(&mut self) -> &mut Self {
+        self.disable_ssl = true;
+
+        self
+    }
+
+    pub fn enable_path_style(&mut self) -> &mut Self {
```

Comment:
not sure if the signature of method `enable_path_style` is composable with the `finish` method, e.g. `builder.enable_path_style().finish()?`

Reply:
even if they are not composable, lets' fix it in #3774

Reply:
Quoted inside https://github.com/datafuselabs/databend/issues/3774